src/serac/physics/materials/tests/thermomechanical_material.cpp

```diff
@@ -0,0 +1,315 @@
+  double free_energy = material.calculateFreeEnergy(displacement_grad, temperature);
+  EXPECT_NEAR(free_energy, 0.0, 1e-10);
+}
+
+TEST(ThermomechanicalMaterial, StressIsZeroInReferenceState)
+{
+  ThermoelasticMaterial material{.E=100.0, .nu=0.25, .C=1.0, .alpha=1.0e-3, .theta_ref=300.0, .k=1.0};
+  tensor<double, 3, 3> displacement_grad{};
+  double temperature = material.theta_ref;
+  tensor<double, 3> temperature_grad{};
+  ThermoelasticMaterial::State state{};
+  auto displacement_grad_old = displacement_grad;
+  double temperature_old = temperature;
+  double dt = 1.0;
+  auto stress = material.calculateMechanicalConstitutiveOutputs(
+      displacement_grad, temperature, temperature_grad, state,
+      displacement_grad_old, temperature_old, dt);
+  EXPECT_NEAR(norm(stress), 0.0, 1e-10);
+}
+
+TEST(ThermomechanicalMaterial, FreeEnergyAndStressAgree)
+{
+  ThermoelasticMaterial material{.E=100.0, .nu=0.25, .C=1.0, .alpha=1.0e-3, .theta_ref=300.0, .k=1.0};
+  tensor<double, 3, 3> displacement_grad{{{0.35490513, 0.60419905, 0.4275843 },
+                                          {0.23061597, 0.6735498 , 0.43953657},
+                                          {0.25099766, 0.27730572, 0.7678207 }}};
+  double temperature = 290.0;
+  tensor<double, 3> temperature_grad{0.87241435, 0.11105156, -0.27708054};
+  ThermoelasticMaterial::State state{};
+  tensor<double, 3, 3> displacement_grad_old{};
+  double temperature_old = temperature;
+  double dt = 1.0;
+  auto energy_and_stress = material.calculateFreeEnergy(
+      make_dual(displacement_grad), temperature);
+  auto stress = material.calculateMechanicalConstitutiveOutputs(
+      displacement_grad, temperature, temperature_grad, state,
+      displacement_grad_old, temperature_old, dt);
+  auto error = stress - get_gradient(energy_and_stress);
+  EXPECT_NEAR(norm(error), 0.0, 1e-12);
+}
+
+TEST(ThermomechanicalMaterial, SatisfiesDissipationInequality)
+{
+  ThermoelasticMaterial material{.E=100.0, .nu=0.25, .C=1.0, .alpha=1.0e-3, .theta_ref=300.0, .k=1.0};
+  tensor<double, 3, 3> displacement_grad{{{0.35490513, 0.60419905, 0.4275843 },
+                                          {0.23061597, 0.6735498 , 0.43953657},
+                                          {0.25099766, 0.27730572, 0.7678207 }}};
+  double temperature = 290.0;
+  tensor<double, 3> temperature_grad{0.87241435, 0.11105156, -0.27708054};
+  double temperature_old = temperature;
+  ThermoelasticMaterial::State state{};
+  tensor<double, 3, 3> displacement_grad_old{};
+  double dt = 1.0;
+  auto generalized_fluxes = material.calculateThermalConstitutiveOutputs(
+      displacement_grad, temperature, temperature_grad, state,
+      displacement_grad_old, temperature_old, dt);
+  auto [heat_capacity, source, heat_flux] = generalized_fluxes;
+  // "inner" didn't work for me, but "dot" does
+  // ask sam and jamie about this
+  auto dissipation = -dot(heat_flux, temperature_grad)/temperature;
```

Comment:
Was this answered?

Reply:
I talked to @samuelpmishLLNL about it. `inner` is intended to work here eventually, but this is not implemented yet. Should I do something to record this elsewhere?

Reply:
Add a TODO and move on imo.  Something like:

```
TODO: change to inner() when implemented
```

Reply:
Brandon and I talked about this in a webex last week. The takeaway was: `inner` _should_ work here too, but we haven't implemented the overloads for the different tensor dimensions yet. 

For some context, those gcc 11.x errors we saw a few months earlier seemed to be caused by a compiler bug that was handling variadic lists of integers incorrectly. Unfortunately, that's the C++ tool that makes it easy to write functions like `inner(A, B)` work for any shaped tensor in 1 function definition, so if we want serac to support gcc 11, then we have to manually write out separate function overloads for the cases 1D tensors, 2D tensors, 3D tensors, etc... -- Oof!